codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -361,9 +362,19 @@ public void error(ChannelHandlerContext ctx, Throwable cause) {
         @Override
         public void write(ChannelHandlerContext ctx, int allowedBytes) {
             int queuedData = queue.readableBytes();
-            if (!endOfStream && (queuedData == 0 || allowedBytes == 0)) {
-                // Nothing to write and we don't have to force a write because of EOS.
-                return;
+            if (!endOfStream) {
+                if (queuedData == 0) {
+                    // There's no need to write any data frames because there are only empty data frames in the queue
+                    // and it is not end of stream yet. Just complete their promises by writing an empty buffer.
+                    ChannelPromise writePromise = ctx.newPromise().addListener(this);
+                    queue.remove(0, writePromise).release();
+                    ctx.write(Unpooled.EMPTY_BUFFER, writePromise);
```

Comment:
IIRC the `EMPTY_BUFFER` won't actually write anything ... this seems more like "just complete the promise after all other writes are complete". This seems to preserve existing behavior where we don't actually write empty frames unless we need to communicate EOS.